### DISCUSSION TITLE
DSL2: Can you emit:  an optional output?

### DISCUSSION THREAD

**Discussion by Mxrcon** [nextflow-io/nextflow#1980] [Q&A] [answered]
Hello there! I'm studying nextflow, specially with  the DSL2 and I'm trying to name an optional output, my code is here:
```
nextflow.enable.dsl=2

process foo{

	output: 
	tuple val(a), file("*.txt")
	tuple val(b), file("*.tsv") optional true, emit: test_channel
	tuple val(c), file("*.csv")
	script:
	a = "1"
	b = "2"
	c = "3"
	"""
	touch nothing.txt
	touch nothing.csv
	"""
}

workflow{
 foo()
}
```

I'm trying to emit and optional output and nextflow returns me this error: 
```
N E X T F L O W  ~  version 21.03.0-edge
Launching `test.nf` [crazy_minsky] - revision: 935e2b6a7d
No such variable: test_channel
```
This feature exists? or should I request it?

Thanks for your attention.

**Comment by mjhipp** [comment]
Hi! I had this issue before, and its actually a really easy fix!
Although it looks strange, you just put `optional true` after the emit.

```nextflow
nextflow.enable.dsl=2

process foo{

	output: 
	tuple val(a), file("*.txt")
	tuple val(b), file("*.tsv"), emit: test_channel optional true // this line is only change
	tuple val(c), file("*.csv")
	script:
	a = "1"
	b = "2"
	c = "3"
	"""
	touch nothing.txt
	touch nothing.csv
	"""
}

workflow{
 foo()
}
```

```
❯ ~/git-repos/nextflow/launch.sh test.nf
N E X T F L O W  ~  version 21.04.0-SNAPSHOT
Launching `test.nf` [sharp_noyce] - revision: 1a2d200173
executor >  local (1)
[64/3094d9] process > foo [100%] 1 of 1 ✔
```

**Reply by Mxrcon** [reply]
Thank you so much! i tried to write in different ways but couldn't find it.

**Comment by pditommaso** [comment, marked as answer]
The correct syntax is 

```
nextflow.enable.dsl=2

process foo{

	output: 
	tuple val(a), file("*.txt")
	tuple val(b), file("*.tsv"), emit: test_channel, optional: true
	tuple val(c), file("*.csv")
	script:
	a = "1"
	b = "2"
	c = "3"
	"""
	touch nothing.txt
	touch nothing.csv
	"""
}

workflow{
 foo()
}
```

**Reply by Mxrcon** [reply]
Thanks @pditommaso for your syntax correction, I'll add your reply as correct answer

**Reply by jonbra** [reply]
Hi, I was searching for this in the documentation, but is this only described for DSL1?
https://www.nextflow.io/docs/latest/process.html:
`output:
    path("output.txt") optional true into outChannel`

**Reply by g1ronn1mo** [reply]
Is it realy a comma behind the emit statement? 
I always get strange results.
